cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -482,14 +482,26 @@ func GetClientURL(localEndpoint *kubeadmapi.APIEndpoint) string {
 	return "https://" + net.JoinHostPort(localEndpoint.AdvertiseAddress, strconv.Itoa(constants.EtcdListenClientPort))
 }
 
+// GetClientURLByIP creates an HTTPS URL based on an IP address
+// and the client listening port.
+func GetClientURLByIP(ip string) string {
+	return "https://" + net.JoinHostPort(ip, strconv.Itoa(constants.EtcdListenClientPort))
+}
+
 // GetPeerURL creates an HTTPS URL that uses the configured advertise
 // address and peer port for the API controller
 func GetPeerURL(localEndpoint *kubeadmapi.APIEndpoint) string {
 	return "https://" + net.JoinHostPort(localEndpoint.AdvertiseAddress, strconv.Itoa(constants.EtcdListenPeerPort))
 }
 
-// GetClientURLByIP creates an HTTPS URL based on an IP address
+// GetMetricsURL creates an HTTP URL that uses the configured advertise
+// address and client port for the API controller
+func GetMetricsURL(localEndpoint *kubeadmapi.APIEndpoint) string {
```

Comment:
to reduce the diff this function could have been added at the bottom bellow GetClientURLByIP

but there are a couple of problems here:
- unsecure listening for metrics outside of localhost is not something we should enable by default. so the bellow returned protocol by this function should be `https://`.
- like @mauilion mentions in the [slack discussion](https://kubernetes.slack.com/archives/CEKK1KTN2/p1589121508249600) exposing the metrics to the public needs to be on port 2379, which possibly suggests setting the value of constants.EtcdMetricsPort to 2379. this change will affect the tests of this function too https://github.com/kubernetes/kubernetes/blob/3c871adc5fe5ce51af436adf94ceaddc666b5a35/cmd/kubeadm/app/util/staticpod/utils.go#L273
as the livenessprobe for the etcd pod will change.

for extra context, the only reason kubeadm added `listen-metrics-urls` is because etcd exposes both the metrics and health endpoints using the same flag, which i argue is a design flaw in etcd.

Reply:
if the client URLs look like this:
>     - --listen-client-urls=https://127.0.0.1:2379,https://10.10.10.2:2379

you already have `https://127.0.0.1:2379/metrics` and `https://10.10.10.2:2379/metrics`, because the client URLs enable that.

if your PR wants to enable insecure serving on e.g. `http://10.10.10.2:2379/metrics`, we cannot do that.

/hold

Reply:
hey @neolit123, I think there is a misunderstanding.  This PR is intending to serve metrics also on **`http://`**`10.10.10.2:2381`.  Before this PR, metrics were already being configured to be served on three different locations: `https://127.0.0.1:2379`, `https://10.10.10.2:2379` and `http://127.0.0.1:2381/metrics`.  So I don't think this is a user facing change.

Reply:
>  I think there is a misunderstanding. This PR is intending to serve metrics also on http://10.10.10.2:2381

yes, this is the part we should not add. only localhost should serve metrics insecurely.

Reply:
what prevents you from using `https://10.10.10.2:2379/metrics` to obtain metrics in your setup?
kubeadm already generates certificates that can be used here.

Reply:
automating using tls is generally clunky and requires additional certs to be mounted into prometheus.  If metrics is already exposed without tls on localhost before this PR, I don't see why we can't also expose it on the service ip?

Reply:
localhost is only accessible for this node, but we don't want a malicious agent looking at the metrics from the same sub-network or even across the internet if the user passes public IPs for their nodes.

you can always patch your etcd.yaml for your particular use case.
this is fairly straight forward and editing ecd.yaml to add the extra URL will just restart etcd fine.

Reply:
Sounds good ma dude, closing this 😄